cpu/native/periph/spi.c

```diff
@@ -0,0 +1,272 @@
+    if (bus >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+
+    mutex_lock(&(device_state[bus].lock));
+
+    bool use_hwcs = false;
+    int fd = -1;
+    if (cs != SPI_CS_UNDEF) {
+        use_hwcs = true;
+        if (cs > SPI_MAXCS || device_state[bus].fd[cs] < 0) {
+            DEBUG("spi_acquire: No fd for %d:%d\n", bus, cs);
+            return SPI_NOCS;
+        }
+        fd = device_state[bus].fd[cs];
+        DEBUG("spi_acquire: Using %d:%d with HWCS (-> fd 0x%x)\n", bus, cs, fd);
+    }
+    else {
+        fd = spidev_get_first_fd(&(device_state[bus]));
+        if (fd < 0) {
+            return SPI_NOCS;
+        }
+        DEBUG("spi_acquire: Using SPI_NO_CS (-> fd 0x%x)\n", fd);
+    }
+
+    int res = spi_set_params(fd, use_hwcs, mode, clk);
+    if (res < 0) {
+        DEBUG("spi_acquire: set_params failed for %d:%d\n", bus, cs);
+        mutex_unlock(&(device_state[bus].lock));
+    }
+
+    DEBUG("spi_acquire: %d:%d acquired\n", bus, cs);
+    return SPI_OK;
+}
+
+void spi_init(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+    spidev_linux_state_t *state = &(device_state[bus]);
+    spidev_linux_conf_t *conf = &(device_conf[bus]);
+
+    spidev_init_device_state(state);
+    DEBUG("spi_init: init bus %d\n", bus);
+    for (spi_cs_t cs = 0; cs < SPI_MAXCS; cs++) {
+        if (conf->device_filename[cs] != NULL) {
+            int fd = real_open(conf->device_filename[cs], O_RDWR);
+            if (fd < 0) {
+                /* Add a printf instead of only asserting to show invalid bus */
+                real_printf(
+                    "Cannot acquire %s for spidev%d:%d\n",
+                    conf->device_filename[cs],
+                    bus,
+                    cs
+                    );
+                assert(false);
+            }
+            DEBUG("spi_init: %d:%d %s (fd 0x%x)\n", bus, cs,
+                  conf->device_filename[cs], fd);
+            state->fd[cs] = fd;
+        }
+        else {
+            DEBUG("spi_init: %d:%d Unused\n", bus, cs);
+        }
+    }
+    DEBUG("spi_init: done\n");
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    if (bus >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+    else if (cs != SPI_CS_UNDEF && cs >= SPI_MAXCS) {
+        return SPI_NOCS;
+    }
+    else if (device_state[bus].fd[cs] < 0) {
+        return SPI_NOCS;
+    }
+    return SPI_OK;
+}
+
+void spi_init_pins(spi_t bus)
+{
+    (void)bus;
+    /* Nothing to do here, as the kernel driver does the pin management */
+}
+
+void spi_release(spi_t bus)
+{
+    DEBUG("spi_release(%d)\n", bus);
+    if (bus < SPI_NUMOF) {
+        mutex_unlock(&(device_state[bus].lock));
+    }
+}
+
+static int spi_set_params(int fd, bool hwcs, spi_mode_t mode, spi_clk_t clk)
+{
+    uint8_t spi_mode = mode | (hwcs ? 0 : SPI_NO_CS);
+
+    if (real_ioctl(fd, SPI_IOC_WR_MODE, &spi_mode) < 0) {
+        return SPI_NOMODE;
+    }
+    uint32_t ioctl_clk = clk;
```

Comment:
for style, but it with the other variable declaration (`uint8_t spi_mode`) above.